File: huggingplace-python-sdk/examples/tracing_llm.py
"""
LLM tracing example for HuggingPlace Python SDK
"""

import asyncio
import os
from huggingplace_sdk import HuggingPlace


async def llm_tracing_example():
    """Example: Trace an LLM call"""
    print('=== LLM Tracing Example ===')
    
    # Initialize the SDK
    hp = HuggingPlace({
        'api_key': 'your-api-key',
        'org_id': 'your-org-id',
        'mode': 'dev',
        'silent': False
    })
    
    # Example LLM function that returns rich response data
    async def mock_openai_call(prompt, **kwargs):
        # Simulate OpenAI API call
        await asyncio.sleep(0.5)  # Simulate API delay
        
        # Return rich response object (like OpenAI would)
        return {
            'responseText': f'This is a response to: {prompt}',
            'rawResponse': {
                'id': 'chatcmpl-123',
                'object': 'chat.completion',
                'created': 1234567890,
                'model': 'gpt-4',
                'choices': [{
                    'index': 0,
                    'message': {
                        'role': 'assistant',
                        'content': f'This is a response to: {prompt}'
                    },
                    'finish_reason': 'stop'
                }],
                'usage': {
                    'prompt_tokens': 10,
                    'completion_tokens': 20,
                    'total_tokens': 30
                }
            },
            'tokens': {
                'input': 10,
                'output': 20,
                'total': 30
            },
            'finishReason': 'stop',
            'model': 'gpt-4',
            'provider': 'openai'
        }
    
    # Trace the LLM call
    result = await hp.trace_llm(
        step_name='OpenAI Chat Completion',
        llm_func=mock_openai_call,
        prompt='What is the weather like today?',
        user_metadata={'user_id': 'user-123'},
        org_data={'org_id': 'org-456'},
        llm_metadata={
            'provider': 'openai',
            'model': 'gpt-4',
            'temperature': 0.7,
            'maxTokens': 100,
            'topP': 1.0,
            'frequencyPenalty': 0.0,
            'presencePenalty': 0.0
        },
        custom_metadata={'session_id': 'session-789'},
        tags=['llm', 'openai', 'chat-completion']
    )
    
    print('LLM tracing result:', result)


async def llm_with_evaluation_example():
    """Example: Trace LLM with evaluation data"""
    print('\n=== LLM with Evaluation Example ===')
    
    hp = HuggingPlace({
        'api_key': 'your-api-key',
        'org_id': 'your-org-id',
        'mode': 'dev'
    })
    
    async def llm_with_evaluation(prompt, **kwargs):
        # Simulate LLM call with evaluation
        await asyncio.sleep(0.3)
        
        response_text = f'Response to: {prompt}'
        
        # Simulate evaluation metrics
        evaluation = {
            'relevance_score': 0.85,
            'fluency_score': 0.92,
            'coherence_score': 0.88,
            'overall_score': 0.88
        }
        
        return {
            'responseText': response_text,
            'rawResponse': {
                'choices': [{'message': {'content': response_text}}],
                'usage': {'total_tokens': 25}
            },
            'tokens': {'total': 25},
            'evaluation': evaluation,
            'model': 'gpt-4',
            'provider': 'openai'
        }
    
    result = await hp.trace_llm(
        step_name='LLM with Evaluation',
        llm_func=llm_with_evaluation,
        prompt='Explain quantum computing in simple terms',
        user_metadata={'user_id': 'user-123'},
        org_data={'org_id': 'org-456'},
        llm_metadata={
            'provider': 'openai',
            'model': 'gpt-4',
            'temperature': 0.5
        },
        custom_metadata={'evaluation_enabled': True},
        tags=['llm', 'evaluation', 'quantum']
    )
    
    print('LLM with evaluation result:', result)


async def run_examples():
    """Run all LLM examples"""
    try:
        await llm_tracing_example()
        await llm_with_evaluation_example()
        
        print('\n=== All LLM examples completed ===')
        print('Check your HuggingPlace backend for LLM trace data!')
    except Exception as error:
        print('LLM example failed:', error)


if __name__ == "__main__":
    asyncio.run(run_examples())

File: huggingplace-sdk/examples/tracing-llm.js
// LLM tracing example for HuggingPlace SDK
import { HuggingPlace, generateId } from '../src/index.js';

// Initialize the SDK with tracing configuration
const hp = new HuggingPlace({
  apiKey: 'your-api-key',
  orgId: 'your-org-id',
  mode: 'dev',
  silent: false,
  traceBatchSize: 10,
  traceBatchTimeout: 5000,
  traceMaxRetries: 3
});

// Example 1: Trace OpenAI API call
async function openAITracingExample() {
  console.log('=== OpenAI Tracing Example ===');
  
  // Simulate OpenAI API call
  const openAIFunction = async () => {
    // In real usage, this would be an actual OpenAI API call
    await new Promise(resolve => setTimeout(resolve, 500));
    
    return {
      responseText: "I'm doing great! How can I help you today?",
      finishReason: "stop",
      inputTokens: 25,
      outputTokens: 15,
      totalTokens: 40,
      rawResponse: {
        choices: [{
          message: { content: "I'm doing great! How can I help you today?" },
          finish_reason: "stop"
        }],
        usage: {
          prompt_tokens: 25,
          completion_tokens: 15,
          total_tokens: 40
        }
      }
    };
  };
  
  const result = await hp.traceLLM({
    prompt: "How are you feeling today?",
    llmFunction: openAIFunction,
    evaluationStatus: "REVIEW",
    user: { userId: "user-123", email: "user@example.com" },
    org: { orgId: "org-456", orgName: "ExampleOrg" },
    llmMetadata: {
      model: "gpt-4",
      provider: "openai",
      temperature: 0.7,
      maxTokens: 100
    },
    customMetadata: {
      appVersion: "1.0.0",
      environment: "development"
    },
    tags: ["openai", "gpt-4", "conversation"],
    sessionId: "session-789"
  });
  
  console.log('OpenAI tracing result:', result);
  
}

// Example 2: Trace Anthropic API call
async function anthropicTracingExample() {
  console.log('\n=== Anthropic Tracing Example ===');
  
  // Simulate Anthropic API call
  const anthropicFunction = async () => {
    await new Promise(resolve => setTimeout(resolve, 300));
    
    return {
      responseText: "I'm an AI assistant designed to help with various tasks. What would you like to know?",
      finishReason: "end_turn",
      inputTokens: 30,
      outputTokens: 20,
      totalTokens: 50,
      rawResponse: {
        content: [{
          type: "text",
          text: "I'm an AI assistant designed to help with various tasks. What would you like to know?"
        }],
        stop_reason: "end_turn",
        usage: {
          input_tokens: 30,
          output_tokens: 20
        }
      }
    };
  };
  
  const result = await hp.traceLLM({
    prompt: "What can you help me with?",
    llmFunction: anthropicFunction,
    evaluationStatus: "APPROVED",
    user: { userId: "user-123" },
    org: { orgId: "org-456" },
    llmMetadata: {
      model: "claude-3-sonnet",
      provider: "anthropic",
      temperature: 0.5,
      maxTokens: 150
    },
    customMetadata: {
      appVersion: "1.0.0"
    },
    tags: ["anthropic", "claude-3", "assistant"]
  });
  
  console.log('Anthropic tracing result:', result);
}

// Example 3: Trace with previous steps context
async function tracingWithContextExample() {
  console.log('\n=== Tracing with Context Example ===');
  
  // Simulate a conversation with context
  const conversationFunction = async () => {
    await new Promise(resolve => setTimeout(resolve, 400));
    
    return {
      responseText: "Based on our previous conversation about AI, I can help you understand machine learning concepts, implement algorithms, or discuss the latest developments in the field.",
      finishReason: "stop",
      inputTokens: 45,
      outputTokens: 35,
      totalTokens: 80,
      rawResponse: {
        choices: [{
          message: { content: "Based on our previous conversation about AI, I can help you understand machine learning concepts, implement algorithms, or discuss the latest developments in the field." },
          finish_reason: "stop"
        }],
        usage: {
          prompt_tokens: 45,
          completion_tokens: 35,
          total_tokens: 80
        }
      }
    };
  };
  
  const previousSteps = [
    { step: "User Introduction", result: "User asked about AI capabilities" },
    { step: "Context Analysis", result: "Identified user's interest in machine learning" }
  ];
  
  const result = await hp.traceLLM({
    prompt: "What specific areas of AI can you help me with?",
    llmFunction: conversationFunction,
    evaluationStatus: "REVIEW",
    user: { userId: "user-123" },
    org: { orgId: "org-456" },
    llmMetadata: {
      model: "gpt-4",
      provider: "openai",
      temperature: 0.3
    },
    previousSteps,
    customMetadata: {
      conversationId: "conv-123",
      turnNumber: 3
    },
    tags: ["conversation", "context-aware", "gpt-4"]
  });
  
  console.log('Context-aware tracing result:', result);
}

// Run all examples
async function runExamples() {
  try {
    await openAITracingExample();
    await anthropicTracingExample();
    await tracingWithContextExample();
    
    console.log('\n=== All LLM tracing examples completed ===');
    console.log('Check your HuggingPlace backend for trace data!');
  } catch (error) {
    console.error('Example failed:', error);
  }
}

// Run if this file is executed directly
// if (import.meta.url === `file://${process.argv[1]}`) {
  runExamples();
// }

export { openAITracingExample, anthropicTracingExample, tracingWithContextExample };
